mgr/dashboard: add hardware status summary

Pegonzal opened the pull request:
On the landing page of the Dashboard, add the hardware status summary

Error:
[screen-capture (19).webm](https://github.com/ceph/ceph/assets/106810838/56f01649-dd32-4414-9d29-11fddc10d8c2)

No errors:
[screen-capture (20).webm](https://github.com/ceph/ceph/assets/106810838/35647317-d060-4c5f-a204-3e27b78ec535)


Fixes:https://tracker.ceph.com/issues/64329





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

nizamial09 commented on src/pybind/mgr/dashboard/controllers/hardware.py:
```diff
@@ -0,0 +1,19 @@
+
+    @Endpoint()
+    @EndpointDoc("Retrieve a summary of the hardware status")
+    def status_summary(self, hostname: Optional[str] = None):
```
```suggestion
    def summary(self, hostname: Optional[str] = None):
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/hardware.service.ts:
```diff
@@ -0,0 +1,15 @@
+
+  constructor(private http: HttpClient) { }
+
+  getStatusSummary(): any {
```
```suggestion
  getSummary(): any {
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.ts:
```diff
@@ -69,6 +70,8 @@ export class DashboardV3Component extends PrometheusListHelper implements OnInit
   telemetryURL = 'https://telemetry-public.ceph.com/';
   origin = window.location.origin;
+  hardwareEnabled: boolean;
+  hardwareSummaryData: any;
```
```suggestion
hardwareSummary: any;
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.ts:
```diff
@@ -146,6 +151,12 @@ export class DashboardV3Component extends PrometheusListHelper implements OnInit
     });
   }
 
+  getHardwareSummaryData() {
```
```suggestion
  getHardwareSummary() {
```

nizamial09 reviewed (commented):
just some initial review. looks progressing! thank you

Pegonzal commented:
Hey @nizamial09 , I updated the PR, could you check it?

nizamial09 commented:
jenkins test make check

nizamial09 commented:
jenkins test windows

nizamial09 commented on src/pybind/mgr/dashboard/controllers/hardware.py:
```diff
@@ -0,0 +1,21 @@
+
+    @Endpoint()
+    @EndpointDoc("Retrieve a summary of the hardware health status")
+    def health(self, categories: Optional[List[str]] = None, hostname: Optional[List[str]] = None):
```
let's call it summary since that's what this endpoint actually provides. and also lets' version these endpoints to EXPERIMENTAL to make sure that this is not fully ready to be consumed.

```suggestion
    def summary(self, categories: Optional[List[str]] = None, hostname: Optional[List[str]] = None):
```

Pegonzal replied:
I think that what we are retriving is a health summary, that's why I went with health, in the end we are checking the health of each hardware item

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -56,7 +56,11 @@
                      summaryType="simplified"
-                     *ngIf="healthData.hosts != null"></cd-card-row>
+                     *ngIf="healthData.hosts != null"
+                     [dropdownEnabled]="hardwareEnabled"
```
i don't think we need this to be sent to the other component. If needed, we can just check the `hardwareHealth`  is empty or not to show the dropdown information
```suggestion
```

Pegonzal replied:
You are right, after my last frontend changes this is definetly not needed, should have gotten rid of it, thanks Nizam

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -56,7 +56,11 @@
+                     *ngIf="healthData.hosts != null"
+                     [dropdownEnabled]="hardwareEnabled"
+                     [dropdownData]="hardwareHealth"
+                     (dropdownError)="setHardwareError($event)">
```
why do we need to emit this from the other component. Since we are already getting the hardware summary in this page, let's count the total error and show the error here instead of emitting from there.

Pegonzal replied:
It is not needed, thanks

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.ts:
```diff
@@ -92,6 +98,9 @@ export class DashboardV3Component extends PrometheusListHelper implements OnInit
     this.interval = this.refreshIntervalService.intervalData$.subscribe(() => {
       this.getHealth();
       this.getCapacityCardData();
+      this.getHardwareConfig().subscribe(() => {
+        this.getHardwareHealth();
+      });
```
a better implementation would be to use the async pipes.

```
    this.hardwareSummary$ = this.hwSubject.pipe(
      switchMap(() => this.hardwareService.getHealth().pipe(
        switchMap((data: any) => {
          this.hasHardwareError = data.host.flawed;
          return of(data)
        })
      ))
    );
    this.isHardwareEnabled$ = this.getHardwareConfig();
    this.interval = this.refreshIntervalService.intervalData$.subscribe(() => {
         ........

      if (this.hardwareEnabled) this.hwSubject.next([]);
    });


  getHardwareConfig(): Observable<any> {
    return this.mgrModuleService.getConfig('cephadm').pipe(
      switchMap((resp: any) => {
        this.hardwareEnabled = resp?.hw_monitoring;
        return of(resp?.hw_monitoring)
      })
    );
  }
   
```
    
    and in html, 

```
    <ng-container *ngIf="isHardwareEnabled$ | async">
        <cd-card-row [data]="healthData.hosts"
                     link="/hosts"
                     title="Host"
                     summaryType="simplified"
                     *ngIf="healthData.hosts != null"
                     [dropdownEnabled]="hardwareEnabled"
                     [dropdownData]="hardwareSummary$ | async">
        </cd-card-row></ng-container>
```

Pegonzal replied:
I had to make couple changes to your suggestion

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.ts:
```diff
@@ -163,4 +180,18 @@ export class DashboardV3Component extends PrometheusListHelper implements OnInit
+    );
+  }
+
+  setHardwareError(error: boolean): void {
+    this.hardwareError = error;
+    this.cdRef.detectChanges();
```
we can remove this and especially the manual change detection since that can cause other issues. let's not do this and store the hardware error like i did in the above comment

Pegonzal replied:
Removed it, thanks.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/api/hardware.service.ts:
```diff
@@ -0,0 +1,17 @@
+  constructor(private http: HttpClient) {}
+
+  getHealth(category: string[] = []): any {
+    return this.http.get<any>(`${this.baseURL}/health/`, {
```
```suggestion
    return this.http.get<any>(`${this.baseURL}/health`, {
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html:
```diff
@@ -169,3 +194,38 @@
+      </div>
+    </li>
+  </ng-container>
+</ng-template>
```
i have made some changes locally which i'll just paste here since that'll be easier
```
diff --git a/src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html b/src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html
index 3a5ef55a14f..a193e2b6b62 100644
--- a/src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html
+++ b/src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html
@@ -1,6 +1,6 @@
 <hr>
 <li class="list-group-item">
-  <div class="d-flex pl-1 pb-2 pt-2">
+  <div class="d-flex pl-1 pb-2 pt-2 position-relative">
     <div class="ms-4 me-auto">
       <a [routerLink]="link"
          *ngIf="link && total > 0; else noLinkTitle"
@@ -12,7 +12,7 @@
         <ng-template ngPluralCase="other">{{ title }}s</ng-template>
       </a>
     </div>
-    <span class="me-3">
+    <span class="me-4">
       <ng-container [ngSwitch]="summaryType">
         <ng-container *ngSwitchCase="'iscsi'">
           <ng-container *ngTemplateOutlet="iscsiSummary"></ng-container>
@@ -28,8 +28,8 @@
         </ng-container>
       </ng-container>
     </span>
-    <span *ngIf="dropdownEnabled"
-          class="me-3">
+    <span *ngIf="dropdownData"
+          class="position-absolute end-0 me-2">
       <a (click)="toggleDropdown()"
          class="dropdown-toggle"
          [attr.aria-expanded]="dropdownToggled"
@@ -196,9 +196,10 @@
 </ng-template>
 
 <ng-template #dropdownTemplate>
-  <ng-container *ngFor="let data of dropdownData.total.category | keyvalue">
-    <li class="list-group-item">
-      <div class="d-flex pl-1 pb-2 pt-2">
+  <ng-container *ngFor="let data of dropdownData?.total.category | keyvalue">
+    <li class="list-group-item"
+        *ngIf="data?.value.error">
+      <div class="d-flex pb-2 pt-2">
         <div class="ms-5 me-auto">
           <span *ngIf="data.value.total"
                 [ngPlural]="data.value.total">
@@ -209,8 +210,7 @@
           </span>
         </div>
         <span *ngIf="data.value.ok"
-              class="me-3">
-          <span>
+              class="me-4">
             {{ data.value.ok }}
             <i class="text-success"
                *ngIf="data.value.ok"
@@ -218,13 +218,12 @@
             </i>
           </span>
           <span *ngIf="data.value.error"
-                class="ms-2">
+                class="me-4 ms-2">
                 {{ data.value.error }}
             <i class="text-danger"
                [ngClass]="[icons.danger]">
             </i>
           </span>
-        </span>
       </div>
     </li>
   </ng-container>
```

nizamial09 replied:
this is what the outcome looks like. you can notice i removed all the items that are green and only shown the item which is in error.  I thought it'll be easier and makes more sense for a user rather than showing everything we have for hardware. Let me know your thought about it.

if we agree on the above then, we don't need to show the drop-down toggle and we can even by default toggle the dropdown when there is an error so user don't need to click to see it.

![image](https://github.com/ceph/ceph/assets/71764184/f710b6c4-6cf9-4b56-94be-e41b029c8a37)

Pegonzal replied:
I don't mind one way or the other but,

on the demo we went the other way around, and neither in [patternfly](https://www.patternfly.org/patterns/dashboard/design-guidelines/) 
nor in [carbon](https://carbondesignsystem.com/patterns/status-indicator-pattern/) I see any reference to omitting the values when everything is fine, rather they use the green tick indicator as well.

Let me know what you think @nizamial09

Pegonzal replied:
Leaving it as it is

nizamial09 replied:
btw you can apply the html changes i made here, because it improves the alignment of the inventory card. Right now it looks a bit broken IMO, could you please apply them?

Pegonzal replied:
addressed

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.ts:
```diff
@@ -19,8 +20,19 @@ export class CardRowComponent implements OnChanges {
+  @Input()
+  dropdownData: any;
+  @Input()
+  dropdownEnabled: boolean = false;
+
+  @Output()
+  dropdownError = new EventEmitter<any>();
```
imo, we can remove these

Pegonzal replied:
Yep, done. Thanks

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.ts:
```diff
@@ -30,5 +42,16 @@ export class CardRowComponent implements OnChanges {
       this.total = this.data;
     }
+
+    if (this.dropdownEnabled && this.dropdownData) {
```
```suggestion
    if (this.dropdownData) {
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.ts:
```diff
@@ -30,5 +42,16 @@ export class CardRowComponent implements OnChanges {
+
+    if (this.dropdownEnabled && this.dropdownData) {
+      if (this.title == 'Host') {
+        this.dropdownTotalError = this.dropdownData.host.flawed;
+        this.dropdownError.emit(this.dropdownTotalError > 0);
```
```suggestion
        this.dropdownTotalError = this.dropdownData.host.flawed;
```

nizamial09 commented on src/pybind/mgr/dashboard/services/orchestrator.py:
```diff
@@ -200,6 +200,13 @@ def stop(self) -> str:
+class HardwareManager(ResourceManager):
+
+    @wait_api_result
+    def common(self, category: str, hostname: Optional[List[str]] = None) -> str:
```
?
```suggestion
    def summary(self, category: str, hostname: Optional[List[str]] = None) -> str:
```

Pegonzal replied:
I left it as `common` since it calls `node_proxy_common` which calls `common`,  the `NodeProxyCache` class already has defined a method called `summary` which I think could end up being confusing calling this as `summary`

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -159,6 +163,16 @@
+                   class="d-flex flex-row">
+                <i class="text-danger"
+                   [ngClass]="[icons.danger, icons.large2x]"></i>
+                <a class="ms-2 mt-n1 lead"
+                   i18n>Hardware</a>
```
```suggestion
                <aspanclass="ms-2 mt-n1 lead"
                   i18n>Hardware</span>
```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -159,6 +163,16 @@
               </div>
             </div>
+
+            <div class="d-flex flex-column ms-5 me-4 mt-4 mb-4">
```
you can use `col` to ensure a proper spacing between `Cluster` and `Hardware` without using spacing. here's how it would look like
```
            <div class="d-flex flex-column ms-4 me-4 mt-4 mb-4">
+            <div class="d-flex flex-column col-md-3 ms-4">
               <ng-template #healthChecks>
                 <ng-container *ngTemplateOutlet="logsLink"></ng-container>
                 <ul>
@@ -164,13 +164,13 @@
               </div>
             </div>
 
-            <div class="d-flex flex-column ms-5 me-4 mt-4 mb-4">
-              <div *ngIf="hardwareEnabled && hardwareError"
+            <div class="d-flex flex-column col-md-3">
+              <div *ngIf="hardwareEnabled && hasHardwareError"
                    class="d-flex flex-row">
                 <i class="text-danger"
                    [ngClass]="[icons.danger, icons.large2x]"></i>
-                <a class="ms-2 mt-n1 lead"
-                   i18n>Hardware</a>
+                <span class="ms-2 mt-n1 lead"
+                   i18n>Hardware</span>
               </div>
             </div>

```

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html:
```diff
@@ -169,3 +194,38 @@
+              {{ data.value.total }}
+            <ng-template ngPluralCase="=0">{{ hwNames[data.key] }}</ng-template>
+            <ng-template ngPluralCase="=1">{{ hwNames[data.key] }}</ng-template>
+            <ng-template ngPluralCase="other">{{ hwNames[data.key] }}s</ng-template>
```
i saw some weird pluralizers like Memorys.. So I think we should do something about it. What about having another pluralize pipe which takes a string and replaces 'y' if its the last char with 'ies' in case its a plural

Pegonzal replied:
Yes, I can do that

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html:
```diff
@@ -169,3 +194,38 @@
+              {{ data.value.total }}
+            <ng-template ngPluralCase="=0">{{ hwNames[data.key] }}</ng-template>
+            <ng-template ngPluralCase="=1">{{ hwNames[data.key] }}</ng-template>
+            <ng-template ngPluralCase="other">{{ hwNames[data.key] }}s</ng-template>
```
?
```suggestion
            <ng-template ngPluralCase="other">{{ hwNames[data.key] | pluralize }}s</ng-template>
```

where pluralize can be a pipe that has something like
```
if (value.endsWith('y')) {
      return value.slice(0, -1) + 'ies';
    } else {
      return value + 's';
    }
 ```

BTW, I haven't tested the above so you'll have to do that.

Pegonzal replied:
Done

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html:
```diff
@@ -169,3 +194,38 @@
+        </div>
+        <span *ngIf="data.value.ok"
+              class="me-3">
+          <span>
```
this looks like an unnecessary span. BTW, there was some styling issue here which I fixed in the patch I put here somewhere

nizamial09 replied:
this is still an issue

Pegonzal replied:
addressed

nizamial09 reviewed (changes requested):
Awesome work here so far @Pegonzal. I left a few comments based on my testing (mostly front-end) and also some minor back-end changes as well. Let me know if you want to discuss something.

nizamial09 commented on src/pybind/mgr/dashboard/services/hardware.py:
```diff
@@ -0,0 +1,91 @@
+
+            for host, items in data.items():
+                count_items = {'total': 0, 'ok': 0, 'error': 0}
+
+                if host not in output['host']:
+                    output['host'][host] = {'flawed': False}
+                for item in items.values():
+                    if item['status']['health'] == 'OK':
+                        count_items['ok'] += 1
+                    else:
+                        output['host'][host]['flawed'] = True
+
+                    count_items['total'] += 1
+                count_items['error'] = count_items['total'] - count_items['ok']
+                # Update items by host's category
+                output['host'][host][category] = count_items
+
+                # Update category total items
+                category_total['total'] += count_items['total']
+                category_total['ok'] += count_items['ok']
+                category_total['error'] += count_items['error']
+
+            if category not in output['total']['category']:
+                output['total']['category'][category] = {}
+            output['total']['category'][category] = category_total
+
+            # Update overall total counts
+            total_count['total'] += category_total['total']
+            total_count['ok'] += category_total['ok']
+            total_count['error'] += category_total['error']
+
+        for host in output['host']:
+            if host == 'flawed':
+                continue
+            if output['host'][host]['flawed']:
+                output['host']['flawed'] += 1
+
+        output['total']['total'] = total_count
+
+        return output
```
i tried to simplify this to something like this
```
category_total = {
            'total': sum(len(items) for items in data.values()),
            'ok': sum(item['status']['health'] == 'OK' for items in data.values() for item in items.values()),
            'error': 0
        }

        for host, items in data.items():
            output['host'].setdefault(host, {'flawed': False})
            output['host'][host]['flawed'] = any(item['status']['health'] != 'OK' for item in items.values())

        category_total['error'] = category_total['total'] - category_total['ok']
        output['total']['category'].setdefault(category, {})
        output['total']['category'][category] = category_total

        total_count['total'] += category_total['total']
        total_count['ok'] += category_total['ok']
        total_count['error'] += category_total['error']

    output['total']['total'] = total_count

    output['host']['flawed'] = sum(1 for host in output['host'] if host != 'flawed' and output['host'][host]['flawed'])

    return output
```

this is also more or less what you did but just taking advantage of the list comprehension.

nizamial09 replied:
it might need testing though, but if its too much to do we can ignore it for now

Pegonzal replied:
Sure, I can make the changes.
I'll use `.setdefault()` to avoid unnecessary comprobations and make use of the comprehensions lists to make the code more concise. 

For future references, I'll try to avoid nesting for loops and use comprehensions list when possible.

Pegonzal commented on src/pybind/mgr/dashboard/services/hardware.py:
```diff
@@ -0,0 +1,91 @@
+
+        orch_hardware_instance = OrchClient.instance().hardware
+        for category in categories:
+            data = json.loads(orch_hardware_instance.common(category, hostname))
```
right, thanks.

nizamial09 replied:
this is unaddressed still

nizamial09 commented:
sorry for getting back here late. I replied to some of the comments here. And I see some of my comments are still unresolved. let me know if something is a blocker for you.

cloudbehl commented:
@Pegonzal commit message `hardware` spelling needs to be fixed.

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/dashboard-v3/dashboard/dashboard-v3.component.html:
```diff
@@ -159,6 +161,16 @@
               </div>
             </div>
+
+            <div class="d-flex flex-column col-md-3">
```
can you apply this same `col-md-3` to Cluster Status as well. col with some initial margin, `col-md-3 ms-4`

nizamial09 reviewed (commented):
Thanks for addressing my comments. I've reopened some of them (mostly html) since I saw they are not addressed.

if the hardware status is enabled, but it doesn't return anything (empty), then we see something like this in the UI. This shouldn't exist.
![Screenshot from 2024-02-29 20-45-36](https://github.com/ceph/ceph/assets/71764184/c8bb0d71-fc95-4720-9cff-1791982e7555)

Pegonzal commented:
Addressed your comments, I think everything is in place now, as per your suggestions @nizamial09 

Adding a video with the updated view 
[status_with_errors.webm](https://github.com/ceph/ceph/assets/106810838/432d746e-9a28-4d51-af53-5144171b8bd1)

cloudbehl commented:
dropdown Icon can be center aligned
![image](https://github.com/ceph/ceph/assets/6695156/de79cc9e-e425-4a80-a1b7-7de005482a87)

Is the space/passing between the icon consistent with how we have on other inventories, to me it looks on higher side.
![Screenshot from 2024-03-01 13-21-21](https://github.com/ceph/ceph/assets/6695156/0880036d-1a53-4cc9-b95d-dc3fcec60678)

![image](https://github.com/ceph/ceph/assets/6695156/5080bea4-0b3a-4136-a649-852092106420)

Pegonzal commented:
> dropdown Icon can be center aligned ![image](https://private-user-images.githubusercontent.com/6695156/309186529-de79cc9e-e425-4a80-a1b7-7de005482a87.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyODIwNTQsIm5iZiI6MTcwOTI4MTc1NCwicGF0aCI6Ii82Njk1MTU2LzMwOTE4NjUyOS1kZTc5Y2M5ZS1lNDI1LTRhODAtYTFiNy03ZGUwMDU0ODJhODcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDFUMDgyOTE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NTdhMWY5MDRkMTRkODdjNDRjNTFkYjcwMGI4MDVjNjJiMzA1YzJhZjcxMjY1MDMyOWQxYWM1YjgzNmRkZGZhNyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.Dm2VihEAY3sW795BRRajPIs4dZFNDlCTfRncQmfRMn0)
> 
> Is the space/passing between the icon consistent with how we have on other inventories, to me it looks on higher side. ![Screenshot from 2024-03-01 13-21-21](https://private-user-images.githubusercontent.com/6695156/309186559-0880036d-1a53-4cc9-b95d-dc3fcec60678.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyODIwNTQsIm5iZiI6MTcwOTI4MTc1NCwicGF0aCI6Ii82Njk1MTU2LzMwOTE4NjU1OS0wODgwMDM2ZC0xYTUzLTRjYzktYjk1ZC1kYzNmY2VjNjA2NzgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDFUMDgyOTE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MmYzOGZiNzBjMGNlYTZiMGVkZjdiYzk4NmNhYmIwY2U1M2VlZDJjMTdmNmRjYTBlZTIwMWViNWFmYzlmMzk5OCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.gYt4FSDP-DNASB6lzQ6vJ3b-WaNcoYoBNxgUam83huk)
> 
> ![image](https://private-user-images.githubusercontent.com/6695156/309187312-5080bea4-0b3a-4136-a649-852092106420.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDkyODIwNTQsIm5iZiI6MTcwOTI4MTc1NCwicGF0aCI6Ii82Njk1MTU2LzMwOTE4NzMxMi01MDgwYmVhNC0wYjNhLTQxMzYtYTY0OS04NTIwOTIxMDY0MjAucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwMSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDFUMDgyOTE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZTVlNzJmMzQ2ZDFjOGM1NGM3ZTk0MDU5Zjk0ZWFmNjk3MmRlMjY2MWJkMTdkMjE3MGI0NzZhM2FhMDM1NTIzMCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.QAz5wb1L4_rL6lvy9AtDoNiszigcL_VHWKL6_vkz-94)

I wasn't able to center it using boostrap classes, so directly added 0.15rem to the margin to align it. Hope that is okay.
For the other's I've conditionally set the margin class, to fit better the spacing as on the other places

![image](https://github.com/ceph/ceph/assets/106810838/3f0e306a-4dd8-4885-8a62-0d5e46468c13)
![image](https://github.com/ceph/ceph/assets/106810838/79707ec3-d1f8-46cd-9c11-1e8e1ef2d187)

nizamial09 commented on src/pybind/mgr/dashboard/frontend/src/app/shared/components/card-row/card-row.component.html:
```diff
@@ -169,3 +194,36 @@
+          </i>
+        </span>
+        <span *ngIf="data.value.error"
+              class="me-4 ms-2">
```
if you remove this `ms-2` and for the above `span`, add a class='me-2', does that work?

Pegonzal replied:
If I understood your suggestion, then you end up with a shorter margin on the right, I don't think it's desired

![image](https://github.com/ceph/ceph/assets/106810838/aa622091-c9a7-4e81-8567-abbafb8b28ed)

nizamial09 reviewed (approved):
LGTM! Thank you so much for the efforts here @Pegonzal

Pegonzal commented:
jenkins test make check

Pegonzal commented:
jenkins test make check
